### PR TITLE
fix: Improvements to date/time handling

### LINKED ISSUE #700: REST: Make date-time and google-datetime format properties call different conversion methods
It looks like these probably have different formats - "date-time" includes a UTC offset, whereas "google-datetime" is always in UTC. Not sure about the precision of either of them - I'd expect "google-datetime" to be nanosecond precision.

### PR DESCRIPTION
**Pull request by jskeet**
This single PR fixes #700 and #701.

It makes #704 and #705 obsolete. (It's basically a rewrite.)

The last commit is the one requiring most concerted review. I'm currently regenerating and rebuilding all packages in google-api-dotnet-client to check for unexpected problems.

### CODE REVIEW

File: Google.Api.Generator.Rest.Tests/GoldenTestData/Google.Apis.Assuredworkloads.v1/Google.Apis.Assuredworkloads.v1.cs [hunk 1]
```diff
@@ -644,13 +644,71 @@ public ListRequest(Google.Apis.Services.IClientService service, string parent) :
...
-                        /// <summary>The end of the time window.</summary>
+                        private object _intervalEndTime;
+
+                        /// <summary>
+                        /// String representation of <see cref="IntervalEndTimeDateTimeOffset"/>, formatted for
+                        /// inclusion in the HTTP request.
+                        /// </summary>
                         [Google.Apis.Util.RequestParameterAttribute("interval.endTime", Google.Apis.Util.RequestParameterType.Query)]
-                        public virtual object IntervalEndTime { get; set; }
+                        public virtual string IntervalEndTimeRaw { get; private set; }
```
Comment (amanda-tarafa) [inline Google.Api.Generator.Rest.Tests/GoldenTestData/Google.Apis.Assuredworkloads.v1/Google.Apis.Assuredworkloads.v1.cs:654]: Thinking more, do we even need this to be public? I like very much to have this property to apply the attribute to, and we control serialization better. But would users need it at all?

**Review by amanda-tarafa — APPROVED** [review]
LGTM, with a question.

File: Google.Api.Generator.Rest.Tests/GoldenTestData/Google.Apis.Assuredworkloads.v1/Google.Apis.Assuredworkloads.v1.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Generator.Rest.Tests/GoldenTestData/Google.Apis.Assuredworkloads.v1/Google.Apis.Assuredworkloads.v1.cs:654]: Users don't, but we fetch this via reflection, and in some cases I can imagine things breaking if we access internal or private properties via reflection in a different assembly. If we're already doing that elsewhere, I'm happy to include it here too - but I suspect we don't.

Comment (jskeet) [reply Google.Api.Generator.Rest.Tests/GoldenTestData/Google.Apis.Assuredworkloads.v1/Google.Apis.Assuredworkloads.v1.cs:654]: Righto. Had you looked enough to be really happy for me to merge? This all feels so fiddly I don't want to merge if you'd like any more time at all. (I'm aware you've LGTM'd, I'm just being super-cautious.)

Comment (amanda-tarafa) [reply Google.Api.Generator.Rest.Tests/GoldenTestData/Google.Apis.Assuredworkloads.v1/Google.Apis.Assuredworkloads.v1.cs:654]: Yep, I think this is fine. I've tried to run over all possible scenarios and I think people that may run into trouble are those that attempt to mix old and new properties. That should be rare, and there's really no reason that I can think so to do so anyway.

Comment (jskeet) [reply Google.Api.Generator.Rest.Tests/GoldenTestData/Google.Apis.Assuredworkloads.v1/Google.Apis.Assuredworkloads.v1.cs:654]: Great, thanks. Merging.

**Pull request merged**